cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,249 @@
+                               artifactId.getVersion(),
+                               scope);
+
+    Long finalLastModifiedTimestamp = lastModifiedTimestamp;
+    File newJarLocation = Retries
+      .callWithRetries(() -> fetchArtifact(artifactId, url, finalLastModifiedTimestamp), retryStrategy);
+
+    // If the lastModifiedTimestamp is null then there is no previous cache to delete and we can just return
+    if (lastModifiedTimestamp == null) {
+      return newJarLocation;
+    }
+
+    // TODO (CDAP-18051): Migrate this cleanup task to its own service
+    // This means we already have a jar but its out of date, we should delete the jar and the unpacked directory
+    File oldJarLocation = getArtifactJarLocation(artifactId, lastModifiedTimestamp);
+    LOG.debug("Got new location as {} with old location as {}", newJarLocation, oldJarLocation);
+    if (!newJarLocation.equals(oldJarLocation)) {
+      File oldUnpackLocation = getUnpackLocalPath(artifactId, lastModifiedTimestamp);
+      LOG.debug("Deleting previously cached jar");
+      try {
+        oldJarLocation.delete();
+        FileUtils.deleteDirectory(oldUnpackLocation);
+      } catch (IOException e) {
+        //Catch and log the exception, this should not cause the operation to fail
+        LOG.warn("Failed to delete old cached jar for artifact {} version {}: {}", artifactId.getArtifact(),
+                 artifactId.getVersion(), e);
+      }
+    }
+    return newJarLocation;
+  }
+
+  /**
+   * Gets the location on the local filesystem for the directory that contains the unpacked artifact. This method
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    File jarLocation = getArtifact(artifactId);
+    File unpackDir = getUnpackLocalPath(artifactId, Long.valueOf(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesnt exist, creating it now");
+      BundleJarUtil.unJar(jarLocation, unpackDir);
+    }
+    return unpackDir;
+  }
+
+  private File fetchArtifact(ArtifactId artifactId, String url,
+                             Long lastModifiedTimestamp) throws IOException, ArtifactNotFoundException {
+    HttpURLConnection urlConn = remoteClient.openConnection(HttpMethod.GET, url);
+    if (lastModifiedTimestamp != null) {
+      LOG.debug("Found existing local version with timestamp {}", lastModifiedTimestamp);
```

Comment:
I'm not sure which part you want me to move. I cant move the if statement because it adds the `if-modified-since` header to the connection. The `openConnection` call needs to stay in this method too for the retries to work.

I've updated the log statement.